File: 프로그래머스/lv2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.java
import java.util.*;

class Solution {
    public String solution(String m, String[] musicinfos) {
      String answer = "";
        m = change(m);

    LinkedHashMap<String, String> hm = new LinkedHashMap<String, String>();
    for(int i=0; i<musicinfos.length; i++){
      String start = musicinfos[i].split(",")[0];
      String end = musicinfos[i].split(",")[1];
      String title = musicinfos[i].split(",")[2];
      String info = musicinfos[i].split(",")[3];
      
      int startTime = Integer.parseInt((start.split(":")[0]))*60 + Integer.parseInt(start.split(":")[1]);
      int endTime = Integer.parseInt((end.split(":")[0]))*60 + Integer.parseInt(end.split(":")[1]);
      int play = endTime-startTime; // 14

      info = change(info);
      int idx = 0;
      String value = "";
      
      while(play!=0){
        if(idx == info.length()) idx = 0;
        value += info.charAt(idx);
        idx++;
        play--;
      }
      System.out.println(value);
      hm.put(title, value);
    }
    int len = 0;
    for(String s : hm.keySet()){
      if(hm.get(s).contains(m) && len < hm.get(s).length()){
        answer = s;
        len = hm.get(s).length();
      }
    }
        return len != 0 ? answer : "(None)";
    }
    
    public String change(String m){
    m = m.replaceAll("C#", "H");
    m = m.replaceAll("D#", "I");
    m = m.replaceAll("F#", "J");
    m = m.replaceAll("G#", "K");
    m = m.replaceAll("A#", "L");
    
    return m;
  }
    
}

File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.js
const convertTimeToMinutes = (time) => {
    const [h, m] = time.split(':');
    return h * 60 + m * 1;
}

function solution(m, musicinfos) {
    const answer = [];
    
    m = m.replaceAll("C#", "Z")
            .replaceAll("D#", "Y")
            .replaceAll("F#", "X")
            .replaceAll("G#", "W")
            .replaceAll("A#", "V")
            .replaceAll("B#", "U");
    
    for (const musicInfo of musicinfos) {
        const [startTime, endTime, title, info] = musicInfo.split(',');
        const diff = convertTimeToMinutes(endTime) - convertTimeToMinutes(startTime);
        const convertInfo = info.replaceAll("C#", "Z")
                                .replaceAll("D#", "Y")
                                .replaceAll("F#", "X")
                                .replaceAll("G#", "W")
                                .replaceAll("A#", "V")
                                .replaceAll("B#", "U");
        
        const newInfo = convertInfo.repeat(Math.ceil(diff / convertInfo.length)).slice(0, diff);
        if (newInfo.includes(m)) answer.push([title, diff]);
    }
    
    if (!answer.length) return "(None)";
    
    answer.sort((a, b) => {
        if (a[1] === b[1]) return 0;
        return b[1] - a[1];
    })
    
    return answer[0][0];
}